contracts/Trader.sol

```diff
@@ -70,31 +78,34 @@ contract Trader is ITrader {
         for (uint256 i = 0; i < n; i++) {
             // retrieve orders and verify their signatures
             // if the order does not exist, it is created here
-            Types.Order storage makeOrder = grabOrder(makers, i, market);
-            Types.Order storage takeOrder = grabOrder(takers, i, market);
+            Perpetuals.Order storage makeOrder = grabOrder(makers, i, market);
+            Perpetuals.Order storage takeOrder = grabOrder(takers, i, market);
 
-            require(makeOrder.targetTracer == market, "TDR: makeOrder market != supplied market");
-            require(takeOrder.targetTracer == market, "TDR: takeOrder market != supplied market");
+            require(makeOrder.market == market, "TDR: makeOrder market != supplied market");
+            require(takeOrder.market == market, "TDR: takeOrder market != supplied market");
 
             address maker = makers[i].order.maker;
             address taker = takers[i].order.maker;
 
+            uint256 makeOrderFilled = filled[Perpetuals.orderId(makeOrder)];
+            uint256 takeOrderFilled = filled[Perpetuals.orderId(takeOrder)];
+
             // calc fill amount
-            uint256 makeRemaining = makeOrder.amount - makeOrder.filled;
-            uint256 takeRemaining = takeOrder.amount - takeOrder.filled;
+            uint256 makeRemaining = makeOrder.amount - makeOrderFilled;
+            uint256 takeRemaining = takeOrder.amount - takeOrderFilled;
             // fill amount is the minimum of order 1 and order 2
             uint256 fillAmount = makeRemaining > takeRemaining ? takeRemaining : makeRemaining;
 
             // match orders
             ITracerPerpetualSwaps(market).matchOrders(makeOrder, takeOrder, fillAmount);
 
             // update order state
-            makeOrder.filled = makeOrder.filled + fillAmount;
-            takeOrder.filled = takeOrder.filled + fillAmount;
+            filled[Perpetuals.orderId(makeOrder)] = makeOrderFilled + fillAmount;
+            filled[Perpetuals.orderId(takeOrder)] = takeOrderFilled + fillAmount;
 
             // increment nonce if filled
-            bool completeMaker = makeOrder.filled == makeOrder.amount;
-            bool completeTaker = takeOrder.filled == takeOrder.amount;
+            bool completeMaker = makeOrderFilled == makeOrder.amount;
```

Comment:
These comparisons are comparing the previous filled amount (before this trade was executed).

They should compare `filled[id] == makeOrder.amount`

Reply:
Consult 2c0820d.